tests.py

```diff
@@ -17,172 +23,183 @@
 torch.manual_seed(42)
 
 
-def test_scaled_dpa():
-    batch_size = 2
-    num_heads = 4
-    seq_length = 5
-    dim_k = 8
+
+BATCH_SIZES = [1, 8, 16, 32]
+NUM_HEADS = [1, 2, 4, 8]
+SEQ_LENGTHS = [10, 50, 100]
+DIM_KS = [8, 16, 32, 64]
+EMB_SIZES = [64, 128, 256, 512]
+DROPOUTS = [0] # can't use other values here since the drop outs occur randomly...
+MAXLEN = [500, 1000, 5000]
+D_FFS = [10, 50, 100, 500]
+NUM_BLOCKS = [1, 2, 4, 8]
+VOCAB_SIZES = [100, 1000, 10000]
+
+@pytest.mark.parametrize("batch_size, num_heads, seq_length, dim_k", list(itertools.product(BATCH_SIZES, NUM_HEADS, SEQ_LENGTHS, DIM_KS)))
```

Comment:
## Requested Change
* Abstract the `list(itertools.product(...))` into a function and then use it in the tests.
* Something along the lines of:
```py
def gen_input(*args):
  return list(itertools.product(*args))

@pytest.mark.parametrize("batch_size, num_heads, seq_length, dim_k", gen_input(BATCH_SIZES, NUM_HEADS, SEQ_LENGTHS, DIM_KS))
def test_scaled_dpa(batch_size, num_heads, seq_length, dim_k):
  ...
```
* Better yet would be to define a decorator for this like:
```py
def with_args(f, *args):
  ...

@with_args(BATCH_SIZES, NUM_HEADS, SEQ_LENGTHS, DIM_KS)
def test_scaled_dpa(batch_size, num_heads, seq_length, dim_k):
  ...
```
but I don't know how to from the top of my head